# Challenge 1 - T-test

In statistics, t-test is used to test if two data samples have a significant difference between their means. There are two types of t-test:

* **Student's t-test** (a.k.a. independent or uncorrelated t-test). This type of t-test is to compare the samples of **two independent populations** (e.g. test scores of students in two different classes). `scipy` provides the [`ttest_ind`](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.ttest_ind.html) method to conduct student's t-test.

* **Paired t-test** (a.k.a. dependent or correlated t-test). This type of t-test is to compare the samples of **the same population** (e.g. scores of different tests of students in the same class). `scipy` provides the [`ttest_re`](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.ttest_rel.html) method to conduct paired t-test.

Both types of t-tests return a number which is called the **p-value**. If p-value is below 0.05, we can confidently declare the null-hypothesis is rejected and the difference is significant. If p-value is between 0.05 and 0.1, we may also declare the null-hypothesis is rejected but we are not highly confident. If p-value is above 0.1 we do not reject the null-hypothesis.

Read more about the t-test in [this article](https://researchbasics.education.uconn.edu/t-test/) and [this Quora](https://www.quora.com/What-is-the-difference-between-a-paired-and-unpaired-t-test). Make sure you understand when to use which type of t-test. 

In [2]:
# Import libraries

import pandas as pd

#### Import dataset

In this challenge we will work on the Pokemon dataset you have used last week. The goal is to test whether different groups of pokemon (e.g. Legendary vs Normal, Generation 1 vs 2, single-type vs dual-type) have different stats (e.g. HP, Attack, Defense, etc.).

In [129]:
# Import dataset

pokemon = pd.read_csv('../data/Pokemon.csv')

pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

#### First we want to define a function with which we can test the means of a feature set of two samples. 

In the next cell you'll see the annotations of the Python function that explains what this function does and its arguments and returned value. This type of annotation is called **docstring** which is a convention used among Python developers. The docstring convention allows developers to write consistent tech documentations for their codes so that others can read. It also allows some websites to automatically parse the docstrings and display user-friendly documentations.

Follow the specifications of the docstring and complete the function.

In [123]:
from scipy.stats import ttest_ind
def t_test_features(s1, s2, features=['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed', 'Total']):
    """Test means of a feature set of two samples
    
    Args:
        s1 (dataframe): sample 1
        s2 (dataframe): sample 2
        features (list): an array of features to test
    
    Returns:
        dict: a dictionary of t-test scores for each feature where the feature name is the key and the p-value is the value
    """
    results = {}

    # Your code here
    
    for feature in features:
        s1_clean = s1[feature].dropna()
        s2_clean = s2[feature].dropna()
        if len(s1_clean) > 1 and len(s2_clean) > 1:
            t_stat, p_value = ttest_ind(s1_clean, s2_clean, equal_var=False)
            results[feature] = p_value
        else:
            results[feature] = 'Not enough data'
       
    
    return results

#### Using the `t_test_features` function, conduct t-test for Lengendary vs non-Legendary pokemons.

*Hint: your output should look like below:*

```
{'HP': 1.0026911708035284e-13,
 'Attack': 2.520372449236646e-16,
 'Defense': 4.8269984949193316e-11,
 'Sp. Atk': 1.5514614112239812e-21,
 'Sp. Def': 2.2949327864052826e-15,
 'Speed': 1.049016311882451e-18,
 'Total': 9.357954335957446e-47}
 ```

In [96]:
# Your code here
legendary_pokemon = pokemon[pokemon['Legendary'] == True]
non_legendary_pokemon = pokemon[pokemon['Legendary'] == False]

t_test_results = t_test_features(legendary_pokemon, non_legendary_pokemon)

print(t_test_results)

{'HP': 1.0026911708035284e-13, 'Attack': 2.520372449236646e-16, 'Defense': 4.8269984949193316e-11, 'Sp. Atk': 1.5514614112239812e-21, 'Sp. Def': 2.2949327864052826e-15, 'Speed': 1.049016311882451e-18, 'Total': 9.357954335957446e-47}


#### From the test results above, what conclusion can you make? Do Legendary and non-Legendary pokemons have significantly different stats on each feature?

#### Your comment here
The features show significant differences between Legendary and non-Legendary Pokemon, as the p-values are all extremely small (much less than 0.05). This suggests that the stats for Legendary Pokemon are significantly different from those of non-Legendary Pokemon in all of the listed features.

#### Next, conduct t-test for Generation 1 and Generation 2 pokemons.

In [101]:
# Your code here
gen1_pokemon = pokemon[pokemon['Generation'] == 1]
gen2_pokemon = pokemon[pokemon['Generation'] == 2]

t_test_2_results = t_test_features(gen1_pokemon, gen2_pokemon)

print(t_test_2_results)

{'HP': 0.14551697834219623, 'Attack': 0.24721958967217725, 'Defense': 0.5677711011725426, 'Sp. Atk': 0.12332165977104388, 'Sp. Def': 0.18829872292645752, 'Speed': 0.00239265937312135, 'Total': 0.5631377907941676}


#### What conclusions can you make?

#### Your comment here
The features in this test show that for most of the features (HP, Attack, Defense, Sp. Atk, Sp. Def, and Total), there is no significant difference between Generation 1 and Generation 2 Pokemon, as their p-values are greater than 0.05. However, for the Speed feature, there is a significant difference (p-value = 0.0024), suggesting that Pokemon from Generation 1 and Generation 2 differ significantly in their Speed stat.

#### Compare pokemons who have single type vs those having two types.

In [106]:
# Your code here
single_type = pokemon[pokemon['Type 2'].isna()]
dual_type = pokemon[pokemon['Type 2'].notna()]

t_test_3_results = t_test_features(single_type, dual_type)

print(t_test_3_results)

{'HP': 0.11314389855379413, 'Attack': 0.00014932578145948305, 'Defense': 2.7978540411514693e-08, 'Sp. Atk': 0.00013876216585667907, 'Sp. Def': 0.00010730610934512779, 'Speed': 0.02421703281819093, 'Total': 1.1157056505229961e-07}


#### What conclusions can you make?

#### Your comment here
The feautres in this t-test results reveal significant differences between Generation 1 and Generation 2 Pokemon in terms of Attack, Defense, Sp. Atk, Sp. Def, and Total stats, with p-values well below 0.05. However, no significant differences were found for HP and Speed, as their p-values exceeded 0.05. This suggests that while the two generations are similar in terms of HP and Speed, they differ notably in their overall combat stats.

#### Now, we want to compare whether there are significant differences of `Attack` vs `Defense`  and  `Sp. Atk` vs `Sp. Def` of all pokemons. Please write your code below.

*Hint: are you comparing different populations or the same population?*

In [135]:
# Import necessary library
from scipy.stats import ttest_rel

# Perform the paired t-test for Attack vs Defense
attack_defense_ttest = ttest_rel(pokemon['Attack'], pokemon['Defense'])

# Perform the paired t-test for Sp. Atk vs Sp. Def
sp_atk_sp_def_ttest = ttest_rel(pokemon['Sp. Atk'], pokemon['Sp. Def'])


print("Attack vs Defense t-test:")
print(f"T-statistic: {attack_defense_ttest.statistic}, P-value: {attack_defense_ttest.pvalue}")

print("Sp. Atk vs Sp. Def t-test:")
print(f"T-statistic: {sp_atk_sp_def_ttest.statistic}, P-value: {sp_atk_sp_def_ttest.pvalue}")

Attack vs Defense t-test:
T-statistic: 4.325566393330478, P-value: 1.7140303479358558e-05
Sp. Atk vs Sp. Def t-test:
T-statistic: 0.853986188453353, P-value: 0.3933685997548122


#### What conclusions can you make?

#### Your comment here
The  paired t-test results reveal that there is a significant difference between the Attack and Defense stats of all Pokémon, with a p-value of 1.71e-05, which is well below the 0.05 threshold. In contrast, the Special Attack and Special Defense stats do not show a significant difference, as indicated by a p-value of 0.39, which is much greater than 0.05. Therefore, we can conclude that while Attack and Defense differ significantly, there is no significant difference between Special Attack and Special Defense.